### PR TITLE
8325674: Constant fold across compares

### PR DESCRIPTION
**Pull request by caojoshua**
For example, `x + 1 < 2` -> `x < 2 - 1` iff we can prove that `x + 1` does not overflow and `2 - 1` does not overflow. We can always fold if it is an `==` or `!=` since overflow will not affect the result of the comparison.

Consider this more practical example:

```
public void foo(int[] arr) {
  for (i = arr.length - 1; i >= 0; --i) {
    blackhole(arr[i]);
  }
}
```

C2 emits a loop guard that looks `arr.length - 1 < 0`. We know `arr.length - 1` does not overflow because `arr.length` is positive. We can fold the comparison into `arr.length < 1`. We have to compute `arr.length - 1` computation if we enter the loop anyway, but we can avoid the subtraction computation if we never enter the loop. I believe the simplification can also help with stronger integer range analysis in https://bugs.openjdk.org/browse/JDK-8275202.

Some additional notes:
* there is various overflow checking code across `src/hotspot/share/opto`. I separated about the functions from convertnode.cpp into `type.hpp`. Maybe the functions belong somewhere else?
* there is a change in Parse::do_if() to repeatedly apply GVN until the test is canonical. We need multiple iterations in the case of `C1 > C2 - X` -> `C2 - X < C1` -> `C2 < X` -> `X > C2`.  This fails the assertion if `BoolTest(btest).is_canonical()`. We can avoid this by applying GVN one more time to get `C2 < X`.
* we should not transform loop backedge conditions. For example, if we have `for (i = 0; i < 10; ++i) {}`, the backedge condition is `i + 1 < 10`. If we transform it into `i < 9`, it messes with CountedLoop's recognition of induction variables and strides.r
* this change optimizes some of the equality checks in `TestUnsignedComparison.java` and breaks the IR checks. I removed those tests.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8325674](https://bugs.openjdk.org/browse/JDK-8325674): Constant fold across compares (**Enhancement** - P5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17853/head:pull/17853` \
`$ git checkout pull/17853`

Update a local copy of the PR: \
`$ git checkout pull/17853` \
`$ git pull https://git.openjdk.org/jdk.git pull/17853/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17853`

View PR using the GUI difftool: \
`$ git pr show -t 17853`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17853.diff">https://git.openjdk.org/jdk/pull/17853.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17853#issuecomment-1944475308)

### CODE REVIEW

**Comment by TobiHartmann** [comment]
I executed some quick testing and `TestConstantFoldCompares::foldAddLtLong` fails with `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation`:

```
1) Compilation of "public boolean compiler.c2.TestConstantFoldCompares.foldAddLtLong(int)":
> Phase "PrintIdeal":
AFTER: print_ideal
  0  Root  === 0 69 70  [[ 0 1 3 23 25 54 56 65 ]] inner 
  1  Con  === 0  [[ ]]  #top
  3  Start  === 3 0  [[ 3 5 6 7 8 9 11 ]]  #{0:control, 1:abIO, 2:memory, 3:rawptr:BotPTR, 4:return_address, 5:compiler/c2/TestConstantFoldCompares:NotNull *, 6:int}
  5  Parm  === 3  [[ 36 ]] Control !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:-1 (line 242)
  6  Parm  === 3  [[ 36 ]] I_O !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:-1 (line 242)
  7  Parm  === 3  [[ 36 ]] Memory  Memory: @BotPTR *+bot, idx=Bot; !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:-1 (line 242)
  8  Parm  === 3  [[ 70 69 36 ]] FramePtr !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:-1 (line 242)
  9  Parm  === 3  [[ 70 69 ]] ReturnAdr !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:-1 (line 242)
 11  Parm  === 3  [[ 24 ]] Parm1: int !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:-1 (line 242)
 23  ConL  === 0  [[ 36 53 ]]  #long:42
 24  ConvI2L  === _ 11  [[ 36 ]]  #long:minint..maxint:www !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:4 (line 242)
 25  ConL  === 0  [[ 36 ]]  #long:1000
 36  CallStaticJava  === 5 6 7 8 1 (24 1 25 1 1 1 23 1 1 1 1 1 ) [[ 37 38 39 41 ]] # Static  java.lang.Math::min long/half ( long, half, long, half ) Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242) !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 37  Proj  === 36  [[ 43 ]] #0 !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 38  Proj  === 36  [[ 70 43 69 48 ]] #1 !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 39  Proj  === 36  [[ 70 69 ]] #2  Memory: @BotPTR *+bot, idx=Bot; !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 41  Proj  === 36  [[ 53 ]] #5 !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 43  Catch  === 37 38  [[ 44 45 ]]  !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 44  CatchProj  === 43  [[ 69 ]] #0@bci -1  !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 45  CatchProj  === 43  [[ 70 48 ]] #1@bci -1  !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 48  CreateEx  === 45 38  [[ 70 ]]  #java/lang/Throwable (java/io/Serializable):NotNull *  Oop:java/lang/Throwable (java/io/Serializable):NotNull * !jvms: Long::min @ bci:2 (line 2031) TestConstantFoldCompares::foldAddLtLong @ bci:8 (line 242)
 53  AddL  === _ 41 23  [[ 58 ]]  !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:11 (line 242)
 54  ConL  === 0  [[ 58 ]]  #long:100
 56  ConI  === 0  [[ 73 ]]  #int:0
 58  CmpL  === _ 53 54  [[ 72 ]]  !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:16 (line 242)
 65  ConI  === 0  [[ 73 ]]  #int:1
 69  Return  === 44 38 39 8 9 returns 73  [[ 0 ]] 
 70  Rethrow  === 45 38 39 8 9 exception 48  [[ 0 ]] 
 72  Bool  === _ 58  [[ 73 ]] [ge]
 73  CMoveI  === _ 72 65 56  [[ 69 ]]  #bool !orig=[71],[67] !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:24 (line 242)

1) Method "public boolean compiler.c2.TestConstantFoldCompares.foldAddLtLong(int)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={"_#ADD#_"}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - failOn: Graph contains forbidden nodes:
         * Constraint 1: "(\\d+(\\s){2}(Add(I|L|F|D|P).*)+(\\s){2}===.*)"
           - Matched forbidden node:
             * 53  AddL  === _ 41 23  [[ 58 ]]  !jvms: TestConstantFoldCompares::foldAddLtLong @ bci:11 (line 242)
```

**Comment by caojoshua** [comment]
I can't exactly repro that test locally, but I get a similar failure in `foldSubLtLong`. The issue comes from `-XX:-TieredCompilation`. I lightly modified the tests and they are passing with or without tiered compilation

**Comment by eme64** [comment]
@caojoshua do you have a benchmark that would help to justify this change?

**Comment by eme64** [comment]
/reviewers 2

Since this is dealing with guards and RangeChecks, we need to review this very carefully.

**Review by eme64 — CHANGES_REQUESTED** [review]
Just did a quick style check before lunch ;)

Not reviewed the logic yet.

File: src/hotspot/share/opto/subnode.cpp [hunk 1]
```diff
@@ -1534,6 +1534,58 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
...
+      const TypeInteger* x_type = phase->type(x)->isa_integer(bt);
+      if (x_type != nullptr) { // x_type can be TOP
+        const TypeInteger* c1_type = phase->type(c1)->is_integer(bt);
+        const TypeInteger* c2_type = phase->type(cmp2)->is_integer(bt);
+        bool sub_no_overflow = !c1_type->can_overflow(cmp1_op, x_type);
+        bool cons_no_overflow = !c1_type->can_overflow(cmp1_op, c2_type);
+        if ((sub_no_overflow && cons_no_overflow && _test.is_less()) ||
+            _test._test == BoolTest::eq || _test._test == BoolTest::ne) {
+          cmp = CmpNode::make(phase->transform(SubNode::make(c1, cmp2, bt)), x,
+                              bt);
```
[the remark this answers is not in the document]
Comment (caojoshua) [reply src/hotspot/share/opto/subnode.cpp]: I added these suggestions. Is there any guidelines for number of columns? My editor auto-formats it to 80 columns following [clangd's defaults](https://llvm.org/docs/CodingStandards.html#source-code-width).

**Comment by caojoshua** [comment]
> @caojoshua do you have a benchmark that would help to justify this change?

I do not. I can turn the tests into micro-benchmarks, but I don't think those simple tests would show any meaningful data. I also think this change is too small to see any difference in larger benchmarks. I think constant folding is a free enough win where we do not have to look for significant differences in benchmarks.

As a followup, this patch exposes the overflow checking APIs, which can be used for https://bugs.openjdk.org/browse/JDK-8325995. This could probably could make a more meaningful difference.

**Comment by eme64** [comment]
> I do not. I can turn the tests into micro-benchmarks, but I don't think those simple tests would show any meaningful data.

I am asking for it because your change brings some risks. Messing with RangeChecks is risky business. Dropping guards wrongly  opens us up to out of bounds access.

> I think constant folding is a free enough win where we do not have to look for significant differences in benchmarks.

Hmm. Not sure this is sufficient. Every optimization brings complexity, bugs and maintenance cost. There are lots of "cool constant folding optimizations", but we cannot do them all.

We already do RangeCheck elimination (RCE) from main-loops. So the overhead is (in most cases) restricted to pre and post loops. Because of RCE, you probably can hardly measure a performance boost with your optimization.

I'm not trying to shoot your work down, but I'm a bit scared of it :sweat_smile: 

Similar questions will apply for your proposal in https://bugs.openjdk.org/browse/JDK-8325995

**Comment by caojoshua** [comment]
> I am asking for it because your change brings some risks. Messing with RangeChecks is risky business. Dropping guards wrongly opens us up to out of bounds access.

Agreed, but if the compiler is smart enough we can make correct optimizations. In this case, I'm using existing APIs which have been working in the real world. I think we can trust it.

> Similar questions will apply for your proposal in https://bugs.openjdk.org/browse/JDK-8325995

@eme64 I left a comment there. Could you take a look? I understand why the benefits of this PR are questionable, but the benefits for that issue should be more apparent. I'd like to know if its a worthwhile direction to work on.

File: src/hotspot/share/opto/subnode.cpp [hunk 2]
```diff
@@ -1534,6 +1534,57 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
...
+        if ((sub_no_overflow && cons_no_overflow && _test.is_less()) ||
+            _test._test == BoolTest::eq || _test._test == BoolTest::ne) {
+          cmp = CmpNode::make(phase->transform(SubNode::make(c1, cmp2, bt)), x, bt);
+          cmp = phase->transform(cmp);
+          BoolNode* bn = new BoolNode(cmp, _test._test);
+          return bn;
+        }
+      }
+    }
+  }
```
Comment (eme64) [inline src/hotspot/share/opto/subnode.cpp:1586]: This looks like heavy code duplication. Can you refactor this? Maybe a helper method?

File: src/hotspot/share/opto/type.cpp [hunk 3]
```diff
@@ -1744,6 +1744,22 @@ const Type *TypeInt::narrow( const Type *old ) const {
...
+  switch (opcode) {
+  case Op_AddI:
+    return add_underflows(_lo, other_int->_lo) ||
+           add_overflows(_hi, other_int->_hi);
+  case Op_SubI:
+    return subtract_underflows(_lo, other_int->_hi) ||
+           subtract_overflows(_hi, other_int->_lo);
+  }
+  return true;
+}
```
Comment (eme64) [inline src/hotspot/share/opto/type.cpp:1761]: Do you maybe want to assert that no other opcode comes in? Or is there a need for non add/sub opcodes to be passed in?

File: src/hotspot/share/opto/subnode.cpp (hunk 2 above)
Comment (caojoshua) [reply src/hotspot/share/opto/subnode.cpp:1586]: I can post a version of this so we can see what it looks like. I actually did this first, but the code got quite ugly.

File: src/hotspot/share/opto/type.cpp (hunk 3 above)
Comment (caojoshua) [reply src/hotspot/share/opto/type.cpp:1763]: Added assertions

File: src/hotspot/share/opto/subnode.cpp (hunk 2 above)
Comment (caojoshua) [reply src/hotspot/share/opto/subnode.cpp:1579]: Pushed changes that combine this. I don't think its better or worse, just different. Open to suggestions.

**Pull request closed without merging**